File: fixtures/sample-repos/playwright/pages/actions/LoginActions.ts
import type { Page } from '@playwright/test';
import { LoginLocators } from '../locators/LoginLocators';

// Action class: receives a locator class, owns user workflows.
// Specs own assertions; actions return state or perform interactions.
export class LoginActions {
  locators: LoginLocators;

  constructor(private page: Page) {
    this.locators = new LoginLocators(page);
  }

  async navigateTo(path: string): Promise<void> {
    await this.page.goto(path, { waitUntil: 'domcontentloaded' });
  }

  async signIn(email: string, password: string): Promise<void> {
    await this.locators.emailInput.fill(email);
    await this.locators.passwordInput.fill(password);
    await this.locators.signInButton.click();
  }
}

File: fixtures/sample-repos/selenium-dotnet/Pages/Actions/LoginActions.cs
using OpenQA.Selenium;
using Sample.SeleniumDotnet.Pages.Locators;

namespace Sample.SeleniumDotnet.Pages.Actions;

public class LoginActions
{
    private readonly LoginLocators _locators;

    public LoginActions(IWebDriver driver)
    {
        _locators = LoginLocators.For(driver);
    }

    public static LoginActions For(IWebDriver driver) => new(driver);

    public void SignIn(string email, string password)
    {
        _locators.EmailInput.SendKeys(email);
        _locators.PasswordInput.SendKeys(password);
        _locators.SignInButton.Click();
    }
}
